File: events/year_2024/day_16/day_16.py
"""
Advent of Code - Year 2024 - Day 16
https://adventofcode.com/2024/day/16
"""

# Standard imports
from collections import deque, defaultdict

# First party imports
from pythonfw.classes import Particule2D


def preprocessing(puzzle_input: str
                  ) -> tuple[set[tuple[int, int]], tuple[int, int], tuple[int, int]]:
    """
    Parses the puzzle input to extract kiosk positions, start, and end coordinates.
    """
    kiosk = set()
    start = (-1, -1)
    end = (-1, -1)

    for y, line in enumerate(puzzle_input.splitlines()):
        for x, c in enumerate(line):
            match c:
                case '#': kiosk.add((x, y))
                case 'S': start = (x, y)
                case 'E': end = (x, y)
                case _: pass

    return kiosk, start, end


def solver(kiosk: set[tuple[int, int]], start: tuple[int, int], end: tuple[int, int]
           ) -> tuple[int, int]:
    """
    Finds the lowest score and the number of unique tiles in the best path from start to end,
    avoiding kiosk positions.
    """
    paths = deque([[(Particule2D(start, (1, 0)), 0,)]])
    scores = defaultdict(lambda: 1_000_000)
    best_paths_tiles = set()
    lowest_score = 1_000_000

    while paths:
        path = paths.popleft()
        pt, path_score = path[-1]

        if pt.xy() == end:
            if path_score == lowest_score:
                best_paths_tiles.update(pt.xy() for (pt, _) in path)
            elif path_score < lowest_score:
                best_paths_tiles = set(pt.xy() for (pt, _) in path)
                lowest_score = path_score
        else:
            for vx, vy in ((pt.vel.x, pt.vel.y),
                           (-pt.vel.y, pt.vel.x),
                           (pt.vel.y, -pt.vel.x)):
                pos_score = scores[(pt.pos.x + vx, pt.pos.y + vy)]

                if path_score < pos_score + 1_000 and (pt.pos.x + vx, pt.pos.y + vy) not in kiosk:
                    delta = 1 if vx == pt.vel.x else 1_001
                    paths.append(path + [(Particule2D((pt.pos.x + vx, pt.pos.y + vy), (vx, vy)),
                                          path_score + delta)])
                    scores[(pt.pos.x + vx, pt.pos.y + vy)] = min(pos_score, path_score + delta)

    return lowest_score, len(best_paths_tiles)

File: events/year_2024/day_16/day_16.rs
aoc::main!();

fn preprocessing(puzzle_input: &str) -> (HashSet<(i32, i32)>, (i32, i32), (i32, i32)) {
    let mut kiosk = HashSet::new();
    let mut start = (0, 0);
    let mut end = (0, 0);

    for (y, line) in puzzle_input.lines().enumerate() {
        for (x, c) in line.chars().enumerate() {
            match c {
                '#' => {
                    kiosk.insert((x as i32, y as i32));
                }
                'S' => {
                    start = (x as i32, y as i32);
                }
                'E' => {
                    end = (x as i32, y as i32);
                }
                _ => {}
            }
        }
    }
    (kiosk, start, end)
}

fn part_1(kiosk: HashSet<(i32, i32)>, start: (i32, i32), end: (i32, i32)) -> i32 {
    let mut paths = VecDeque::new();
    paths.push_back(vec![(start, (1, 0), 0)]);
    let mut scores = HashMap::new();
    let mut best_paths_tiles = HashSet::new();
    let mut lowest_score = 1_000_000;

    while let Some(path) = paths.pop_front() {
        let ((x, y), (dx, dy), path_score) = path.last().unwrap();
        if (*x, *y) == end {
            if *path_score == lowest_score {
                for (pos, _, _) in &path {
                    best_paths_tiles.insert(*pos);
                }
            } else if *path_score < lowest_score {
                best_paths_tiles.clear();
                for (pos, _, _) in &path {
                    best_paths_tiles.insert(*pos);
                }
                lowest_score = *path_score;
            }
        } else {
            for (vx, vy) in [(dx, dy), (-dy, dx), (dy, -dx)] {
                let pos = (x + vx, y + vy);
                let pos_score = *scores.get(&pos).unwrap_or(&1_000_000);
                if *path_score < pos_score + 1_000 && !kiosk.contains(&pos) {
                    let delta = if vx == *dx { 1 } else { 1_001 };
                    let mut new_path = path.clone();
                    new_path.push((pos, (vx, vy), path_score + delta));
                    paths.push_back(new_path);
                    scores.insert(pos, pos_score.min(path_score + delta));
                }
            }
        }
    }
    lowest_score
}

fn part_2(kiosk: HashSet<(i32, i32)>, start: (i32, i32), end: (i32, i32)) -> usize {
    let mut paths = VecDeque::new();
    paths.push_back(vec![(start, (1, 0), 0)]);
    let mut scores = HashMap::new();
    let mut best_paths_tiles = HashSet::new();
    let mut lowest_score = 1_000_000;

    while let Some(path) = paths.pop_front() {
        let ((x, y), (dx, dy), path_score) = path.last().unwrap();
        if (*x, *y) == end {
            if *path_score == lowest_score {
                for (pos, _, _) in &path {
                    best_paths_tiles.insert(*pos);
                }
            } else if *path_score < lowest_score {
                best_paths_tiles.clear();
                for (pos, _, _) in &path {
                    best_paths_tiles.insert(*pos);
                }
                lowest_score = *path_score;
            }
        } else {
            for (vx, vy) in [(dx, dy), (-dy, dx), (dy, -dx)] {
                let pos = (x + vx, y + vy);
                let pos_score = *scores.get(&pos).unwrap_or(&1_000_000);
                if *path_score < pos_score + 1_000 && !kiosk.contains(&pos) {
                    let delta = if vx == *dx { 1 } else { 1_001 };
                    let mut new_path = path.clone();
                    new_path.push((pos, (vx, vy), path_score + delta));
                    paths.push_back(new_path);
                    scores.insert(pos, pos_score.min(path_score + delta));
                }
            }
        }
    }
    best_paths_tiles.len()
}
